File: src/GnOuGo.Flow.Core/Runtime/Executors/WorkflowCallExecutor.cs
using System.Text.Json.Nodes;
using Microsoft.Extensions.Logging;
using GnOuGo.Flow.Core.Expressions;
using GnOuGo.Flow.Core.Models;

namespace GnOuGo.Flow.Core.Runtime.Executors;

/// <summary>
/// Calls a local or remote sub-workflow.
/// </summary>
public sealed class WorkflowCallExecutor : IStepExecutor
{
    public string StepType => "workflow.call";

    public IReadOnlyList<StepExceptionDoc>? DocumentedExceptions => new StepExceptionDoc[]
    {
        new(ErrorCodes.InputValidation, false, "The workflow reference is malformed, the target name/url is missing, or the target workflow cannot be resolved."),
        new(ErrorCodes.WorkflowCycleDetected, false, "The call depth limit was exceeded or a local workflow cycle was detected."),
        new(ErrorCodes.WorkflowFetchPolicy, false, "A remote workflow violates HTTPS/host/export policy constraints."),
        new(ErrorCodes.WorkflowFetchNetwork, false, "A remote workflow fetcher is missing or the remote workflow could not be fetched."),
        new(ErrorCodes.WorkflowFetchIntegrity, false, "The remote workflow integrity check failed.")
    };

    public string DslSnippet => """
        ### workflow.call — Call a sub-workflow (local or remote)
        ```yaml
        - id: sub_task
          type: workflow.call
          input:
            ref:
              kind: local                       # "local" (same document) or "url" (remote)
              name: helper_workflow              # workflow name (local) or URL (remote)
            args:                               # input arguments for the sub-workflow
              key: "${data.inputs.value}"
        ```
        Output: `{ outputs: { ... }, workflow: "name", run: { steps_executed, success } }`
        """;

    public async Task<JsonNode?> ExecuteAsync(StepExecutionContext ctx, CancellationToken ct)
    {
        var input = ctx.Engine.GetResolvedInput(ctx) as JsonObject
            ?? throw new WorkflowRuntimeException(ErrorCodes.InputValidation, "workflow.call input must be object");

        var refObj = input["ref"] as JsonObject
            ?? throw new WorkflowRuntimeException(ErrorCodes.InputValidation, "workflow.call requires 'ref'");

        var kind = refObj["kind"]?.GetValue<string>() ?? "local";
        var args = input["args"] ?? new JsonObject();

        // Check call depth
        if (ctx.CallDepth >= ctx.Limits.MaxCallDepth)
        {
            ctx.Engine.Logger.LogError("workflow.call: max call depth ({MaxCallDepth}) exceeded", ctx.Limits.MaxCallDepth);
            throw new WorkflowRuntimeException(ErrorCodes.WorkflowCycleDetected,
                $"Max call depth ({ctx.Limits.MaxCallDepth}) exceeded");
        }

        var resolution = await ctx.Engine.WorkflowCallResolver.ResolveAsync(new WorkflowCallResolutionContext
        {
            Engine = ctx.Engine,
            Ref = refObj,
            Kind = kind,
            CallDepth = ctx.CallDepth,
            CallStack = ctx.CallStack
        }, ct);

        if (!string.IsNullOrWhiteSpace(resolution.CallStackKey) && ctx.CallStack.Contains(resolution.CallStackKey))
        {
            ctx.Engine.Logger.LogError("workflow.call: cycle detected, workflow call '{WorkflowCallKey}' already in call stack", resolution.CallStackKey);
            throw new WorkflowRuntimeException(ErrorCodes.WorkflowCycleDetected,
                $"Cycle detected: workflow '{resolution.WorkflowName}' already in call stack");
        }

        return await ExecuteResolvedWorkflow(ctx, resolution, args, ct);
    }

    private static async Task<JsonNode?> ExecuteResolvedWorkflow(
        StepExecutionContext ctx,
        WorkflowCallResolution resolution,
        JsonNode? args,
        CancellationToken ct)
    {
        var subWorkflow = resolution.Workflow;
        var newCallStack = new HashSet<string>(ctx.CallStack);
        if (!string.IsNullOrWhiteSpace(resolution.CallStackKey))
            newCallStack.Add(resolution.CallStackKey);

        var resolvedArgs = WorkflowInputDefaults.Apply(subWorkflow.Source, args);
        var inputErrors = InputTypeValidator.Validate(subWorkflow.Source, resolvedArgs);
        if (inputErrors.Count > 0)
        {
            throw new WorkflowRuntimeException(
                ErrorCodes.InputValidation,
                $"Input validation failed for called workflow '{resolution.WorkflowName}': {string.Join("; ", inputErrors)}",
                details: new JsonObject
                {
                    ["workflow"] = resolution.WorkflowName,
                    ["validation_errors"] = new JsonArray(inputErrors.Select(static error => (JsonNode)JsonValue.Create(error)!).ToArray())
                });
        }

        var result = new RunResult { Success = true };
        var subData = new JsonObject
        {
            ["inputs"] = resolvedArgs,
            ["steps"] = new JsonObject(),
            ["env"] = ctx.Data["env"]?.DeepClone() ?? new JsonObject()
        };

        var previousDocument = ctx.Engine.ReplaceCompiledDocumentForWorkflowCall(subWorkflow.Document);
        try
        {
            await ctx.Engine.ExecuteStepsAsync(subWorkflow.Steps, subData, result, ctx.Limits, ctx.CallDepth + 1, newCallStack, ct, ctx.TelemetrySpan);
        }
        finally
        {
            ctx.Engine.ReplaceCompiledDocumentForWorkflowCall(previousDocument);
        }

        // Evaluate outputs
        JsonNode? outputs;
        if (subWorkflow.Outputs != null)
        {
            var outputObj = new JsonObject();
            foreach (var kv in subWorkflow.Outputs)
            {
                outputObj[kv.Key] = ctx.Engine.EvaluateOutputDef(kv.Value, subData);
            }
            outputs = outputObj;
        }
        else
        {
            outputs = subData["steps"]?.DeepClone();
        }

        return new JsonObject
        {
            ["outputs"] = outputs,
            ["workflow"] = resolution.WorkflowName
        };
    }
}

File: clients/python/gnougo-flow-core/src/gnougo_flow_core/runtime_steps/workflow_call_executor.py
from __future__ import annotations

from gnougo_flow_core.runtime import *  # noqa: F401,F403

class WorkflowCallExecutor:
    step_type = "workflow.call"
    step_description = "Call a local or remote workflow by reference."
    dsl_snippet = """
### workflow.call - Execute another workflow
```yaml
- id: run_sub
  type: workflow.call
  input:
    ref:
      kind: local
      name: generated
    args:
      task: "${data.inputs.task}"
```
Output: `{ outputs: <workflow outputs>, workflow: <name> }`.
"""
    documented_exceptions = [
        (ErrorCodes.WORKFLOW_CYCLE_DETECTED, False, "recursive cycle or max depth exceeded."),
        (ErrorCodes.WORKFLOW_FETCH_POLICY, False, "remote workflow reference violates fetch policy."),
        (ErrorCodes.WORKFLOW_FETCH_NETWORK, False, "failed to fetch remote workflow."),
    ]

    async def execute_async(self, ctx: StepExecutionContext) -> Any:
        input_obj = ctx.engine.get_resolved_input(ctx)
        if not isinstance(input_obj, dict):
            raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, "workflow.call input must be object")

        ref = input_obj.get("ref")
        if not isinstance(ref, dict):
            raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, "workflow.call requires 'ref'")

        kind = str(ref.get("kind", "local"))
        args = input_obj.get("args") or {}

        if ctx.call_depth >= ctx.limits.max_call_depth:
            raise WorkflowRuntimeException(ErrorCodes.WORKFLOW_CYCLE_DETECTED, "Max call depth exceeded")

        if kind == "local":
            name = ref.get("name")
            if not isinstance(name, str):
                raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, "Local workflow.call requires 'name'")
            if name in ctx.call_stack:
                raise WorkflowRuntimeException(ErrorCodes.WORKFLOW_CYCLE_DETECTED, f"Cycle detected: workflow '{name}' already in call stack")

            compiled_doc = ctx.engine.compiled_document
            if not compiled_doc or name not in compiled_doc.workflows:
                raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, f"Local workflow '{name}' not found")
            sub = compiled_doc.workflows[name]
            sub_data = {"inputs": dict(args), "steps": {}, "env": dict(ctx.data.get("env", {}))}
            rr = RunResult(success=True)
            await ctx.engine.execute_steps_async(sub.steps, sub_data, rr, ctx.limits, ctx.call_depth + 1, set(ctx.call_stack) | {name}, ctx.telemetry_span)
            outputs = {k: ctx.engine.evaluate_output_def(v, sub_data) for k, v in (sub.outputs or {}).items()} if sub.outputs else sub_data.get("steps", {})
            return {"outputs": outputs, "workflow": name}

        if kind == "url":
            fetcher = ctx.engine.workflow_fetcher
            if fetcher is None:
                raise WorkflowRuntimeException(ErrorCodes.WORKFLOW_FETCH_NETWORK, "No workflow fetcher configured")
            url = ref.get("url")
            if not isinstance(url, str):
                raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, "Remote workflow.call requires 'url'")

            if ctx.engine.fetch_policy and ctx.engine.fetch_policy.require_https and not url.startswith("https://"):
                raise WorkflowRuntimeException(ErrorCodes.WORKFLOW_FETCH_POLICY, "HTTPS required by policy")

            yaml_text = await fetcher.fetch_async(url, ref.get("integrity"))
            doc = WorkflowParser.parse(yaml_text)
            compiled = WorkflowCompiler().compile(doc)
            wf = compiled.workflows.get(compiled.entrypoint or "") or next(iter(compiled.workflows.values()))
            sub_data = {"inputs": dict(args), "steps": {}, "env": dict(ctx.data.get("env", {}))}
            rr = RunResult(success=True)
            await ctx.engine.execute_steps_async(wf.steps, sub_data, rr, ctx.limits, ctx.call_depth + 1, set(ctx.call_stack), ctx.telemetry_span)
            return {"outputs": sub_data.get("steps", {}), "workflow": wf.name}

        raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, f"Unknown workflow.call kind: {kind}")
